### PR TITLE
[ETH-FLOW V2] - Contract hooks and flow functions

### PR DESCRIPTION
**Pull request by W3stside**
# Note
Don't review this, i'll re-open

# Summary

PR adds contract abi and flow functions (based on `swap/services/swapFlow`)

- add ethflow contract abi and hook
- update `contracts_add_export.js` to account for eth-flow contracts
- add `ethFlow.ts` logic similar to `swap/services/swapFlow.ts`
- updates dispatch actions to differentiate between normal off-chain orders and eth-flow orders
- moves common steps files to common folder

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -124,7 +124,7 @@
     "serve": "^11.3.2",
     "start-server-and-test": "^1.11.0",
     "text-encoding-polyfill": "^0.6.7",
-    "typechain": "^5.0.0",
+    "typechain": "^8.1.0",
```
Comment (W3stside) [inline package.json:127]: updating typechain to make ethflow contract compilation work

File: scripts/contracts_add_export.js [hunk 2]
```diff
@@ -4,13 +4,22 @@
 const fs = require('fs')
 const path = require('path')
 
-const file = path.resolve(__dirname, '../src/cow-react/abis/types/index.ts')
-const output = 'export * from "@src/abis/types";\n'
+const fileTypesIndex = path.resolve(__dirname, '../src/cow-react/abis/types/index.ts')
+const outputBaseAbis = 'export * from "@src/abis/types";\n'
+const outputEthFlowAbis = 'export * from "@cow/abis/types/ethflow";\n'
```
Comment (W3stside) [inline scripts/contracts_add_export.js:9]: export ethflow contracts

File: src/cow-react/modules/swap/containers/SwapButtons/index.tsx [hunk 3]
```diff
@@ -64,13 +64,6 @@ const swapButtonStateMap: { [key in SwapButtonState]: (props: SwapButtonsContext
       <Trans>Unwrap</Trans>
     </ButtonPrimary>
   ),
-  [SwapButtonState.SwapWithWrappedToken]: (props: SwapButtonsContext) => (
```
Comment (W3stside) [inline src/cow-react/modules/swap/containers/SwapButtons/index.tsx:67]: we no longer need this state

File: src/cow-react/modules/swap/hooks/useFlowContext.ts [hunk 4]
```diff
@@ -71,20 +102,112 @@ export function useSwapFlowContext(): SwapFlowContext | null {
...
+  contract: GPv2Settlement | null
+  addOrderCallback: AddOrderCallback<AddUnserialisedPendingOrderParams>
+}
+
+interface GetEthFlowContextProps extends BaseGetFlowContextProps {
+  contract: CoWSwapEthFlow | null
+  addOrderCallback: AddOrderCallback<AddUnserialisedPendingEthFlowOrderParams>
+}
+
+function getFlowContext({
```
Comment (W3stside) [inline src/cow-react/modules/swap/hooks/useFlowContext.ts:157]: function overloads to reuse logic and use a generic flow context function

File: src/cow-react/modules/swap/services/ethFlow/index.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import { PriceImpact } from 'hooks/usePriceImpact'
+import { addPendingEthFLowOrderStep } from '@cow/modules/swap/services/ethFlow/steps/addPendingEthFLowOrderStep'
+
+export async function ethFlow(input: EthFlowContext, priceImpactParams: PriceImpact) {
+  logSwapFlow('ETH FLOW', 'STEP 1: confirm price impact')
+  if (priceImpactParams?.priceImpact && !confirmPriceImpactWithoutFee(priceImpactParams.priceImpact)) return
+
+  logSwapFlow('ETH FLOW', 'STEP 2: send transaction')
+  // TODO: check
+  swapFlowAnalytics.swap(input.swapFlowAnalyticsContext)
```
Comment (W3stside) [inline src/cow-react/modules/swap/services/ethFlow/index.ts:16]: should we use the same swapFlow analytics or create a unique one for eth-flow?

File: src/cow-react/modules/swap/services/ethFlow/index.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    logSwapFlow('ETH FLOW', 'STEP 4: add pending order step')
+    addPendingEthFLowOrderStep(
+      {
+        id: orderId,
+        chainId: input.context.chainId,
+        order: {
+          ...order,
+          // TODO: review
+          presignGnosisSafeTxHash: undefined,
```
Comment (W3stside) [inline src/cow-react/modules/swap/services/ethFlow/index.ts:33]: brain hurting not sure if safe presign is applicable for eth-flow orders @josojo

File: src/cow-react/modules/swap/services/ethFlow/index.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      },
+      input.dispatch
+    )
+
+    logSwapFlow('ETH FLOW', 'STEP 5: add app data to upload queue')
+    input.callbacks.addAppDataToUploadQueue({ chainId: input.context.chainId, orderId, appData: input.appDataInfo })
+
+    logSwapFlow('ETH FLOW', 'STEP 6: show UI of the successfully sent transaction', orderId)
+    input.swapConfirmManager.transactionSent(orderId)
+    swapFlowAnalytics.sign(input.swapFlowAnalyticsContext)
```
Comment (W3stside) [inline src/cow-react/modules/swap/services/ethFlow/index.ts:44]: should we create a unique eth-flow analytics or just make the swapFlowAnyltics more generic and accept a prop/flag to signify eth-flow order?

File: src/cow-react/modules/swap/services/ethFlow/index.ts (hunk 5 above)
Comment (josojo) [reply src/cow-react/modules/swap/services/ethFlow/index.ts:16]: @MindyCoW  would know best

File: src/cow-react/modules/swap/services/ethFlow/index.ts (hunk 6 above)
Comment (josojo) [reply src/cow-react/modules/swap/services/ethFlow/index.ts:33]: Gnosis safes don't need any presign. They can just use the normal onchain order placement for the ethflow. They only needed pre_sign for normal orders, as they can not provide any normal signatures

File: .env [hunk 8]
```diff
@@ -55,33 +56,33 @@ REACT_APP_PATH_REGEX_ENS="/ipfs"
...
 # API
-#REACT_APP_API_URL_PROD_MAINNET=https://protocol-mainnet.gnosis.io/api
-#REACT_APP_API_URL_PROD_RINKEBY=https://protocol-rinkeby.gnosis.io/api
-#REACT_APP_API_URL_PROD_XDAI=https://protocol-xdai.gnosis.io/api
-#REACT_APP_API_URL_PROD_GOERLI=https://protocol-goerli.gnosis.io/api
-#REACT_APP_API_URL_STAGING_MAINNET=https://protocol-mainnet.dev.gnosisdev.com/api
-#REACT_APP_API_URL_STAGING_RINKEBY=https://protocol-rinkeby.dev.gnosisdev.com/api
-#REACT_APP_API_URL_STAGING_XDAI=https://protocol-xdai.dev.gnosisdev.com/api
-#REACT_APP_API_URL_STAGING_GOERLI=https://protocol-goerli.dev.gnosisdev.com/api
+#REACT_APP_API_URL_PROD_MAINNET=https://api.cow.fi/mainnet/api
```
Comment (josojo) [inline .env:59]: Maybe we can do the switch in another PR?

File: src/cow-react/abis/types/factories/MerkleDrop__factory.ts [hunk 9]
```diff
@@ -2,42 +2,45 @@
...
+import type { MerkleDrop, MerkleDropInterface } from "../MerkleDrop";
 
 const _abi = [
   {
     inputs: [
       {
-        internalType: 'uint256',
-        name: 'index',
-        type: 'uint256',
+        internalType: "uint256",
```
Comment (josojo) [inline src/cow-react/abis/types/factories/MerkleDrop__factory.ts:13]: I think these are not intended changes.

**Comment by W3stside** [comment]
closing in favour of broken up #1235 and #1241
